### ISSUE TITLE
Explicitly providing sequence length is not ideal

### ISSUE THREAD

**Issue by fornasierov** [mnchl-kim/GRU-D-TF2#2]
Hello, dear @mnchl-kim. I'm trying to use this implementation for data with varying sequence lengths and this is not permitted by the current code. 

My data is of varying lengths (eg., 89 days of sampling for subject A and 23 days of sampling for subject B). I'm using a personalized generator that produces batches of subjects with same number of sampling days. Because of that, I cannot explicitly provide the sequence lengths when building the model.

Bellow is the portion of the code I'm referring to, from [models.py](https://github.com/mnchl-kim/GRU-D-TF2/blob/main/models.py):

    class GRUD(Model):
        def __init__(self, hid_dim, dropout_rate, name='GRU-D', **kwargs):
            super(GRUD, self).__init__(name=name, **kwargs)
    
            self.hid_dim = hid_dim
            self.dropout_rate = dropout_rate
            self.inputs_mean = 0
    
        def build(self, input_shape):
            self.seq_len = input_shape[2] # Right here!
            self.var_dim = input_shape[3]

The seq_len is later used here:

    for t in range(self.seq_len):
        x = inputs[:, 0, t, :]
        m = inputs[:, 1, t, :]
        d = inputs[:, 2, t, :]
        f = inputs[:, 3, t, :]

Is there a way to let _seq_len_ be adjustable according to the _seq_len_ in a given batch (e.g., using **None** when building the model)?

**Comment by mnchl-kim** [comment]
Hmm,,
If you produced batches of subjects with same number of sampling days, then isn't `self.seq_len`  fixed at least in given batches?
If it is true, I think there is no problem,,,,

Instead of using `self.seq_len = input_shape[2]` in `build` method, try using `self.seq_len = tf.shape(inputs)[2]` in `call` method.
`tf.shape` can represent **None** type, which is for dynamic inputs.

However, although `tf.shape` can represent **None** type, for loop can't take **None** type....

**Comment by fornasierov** [comment]
Yes, indeed the for loop cannot work with None type. I believe the whole call function must change to [tf.while_loop](https://www.tensorflow.org/api_docs/python/tf/while_loop) in order to accommodate dynamic shapes. As suggest by you, I've tried using `tf.shape` and it worked alright, but then the problem became a change in shape for `state_h`. See the error:

` ValueError: 'state_h' has shape (16,) before the loop, but shape (None, 16) after one iteration. Use tf.autograph.experimental.set_loop_options to set shape invariants.`

I am not a developer, but I will try to solve this here. I am curious to know if you had the same problem at some point. If so, could you share it in here so I can use as starting point to modify the code for my specific data set?

**Comment by mnchl-kim** [comment]
I had met similar problem before, but I had failed to solve the problem by handling **None** type data structure with tf2....

I had tried 3 methods,,,
- tf.while_loop with **None** type
- PyTorch
- fixed seq_len with zero padding

After trying all above methods, I chose fixed seq_len with zero padding, which is a common alternative technique for time series data.
I hope it helps you. Thank you.
